sw/cheri/common/sonata_plic.hh

```diff
@@ -0,0 +1,185 @@
+    I2c1ReceiveFifoOverflow = 35,
+    I2c1ReceiveNack = 36,
+    I2c1SclInterference = 37,
+    I2c1SdaInterference = 38,
+    I2c1StretchTimeout = 39,
+    I2c1SdaUnstable = 40,
+    I2c1CommandComplete = 41,
+    I2c1TransmitStretch = 42,
+    I2c1TransmitThreshold = 43,
+    I2c1AcquiteFifoFull = 44,
+    I2c1UnexpectedStop = 45,
+    I2c1HostTimeout = 46,
+    // Ethernet Interrupts.
+    Ethernet = 47,
+    MaxIntrID,
+ }Interrupts;
+
+ enum class Priority{
+    PriorityHigh,
+    PriorityMedium,
+    PriorityLow,
+  };
+
+  /**
+   * Driver for the standard RISC-V Platform-Local Interrupt Controller (PLIC).
+   */
+  class SonataPlic {
+    private:
+      // generic offsets according to spec
+      static constexpr size_t PriorityOffset  = 0x0U;
+      static constexpr size_t PendingOffset   = 0x1000U;
+      static constexpr size_t EnableOffset    = 0x2000U;
+      static constexpr size_t ThresholdOffset = 0x200000U;
+      static constexpr size_t ClaimOffset     = 0x200004U;
+
+      // Bounded capabilities to the individual structure fields.
+      // Ideally these should be contained in one volatile struct, but
+      // they are so far apart (and this lets us apply the bounds once).
+      volatile uint32_t *plicPrios;
+      volatile uint32_t *plicPendings;
+      volatile uint32_t *plicEnables;
+      volatile uint32_t *plicThres;
+      volatile uint32_t *plicClaim;
+
+    public:
+      /**
+       * Constructor. Initialises the interrupt controller with all interrupts disabled.
+       */
+      SonataPlic(CHERI::Capability<void> root ){
+        constexpr auto maxIntrID = static_cast<uint32_t>(Interrupts::MaxIntrID);
+        constexpr uint32_t baseAddr = PLIC_ADDRESS;
+        size_t             nSources = (maxIntrID + 32U) & ~0x1f;
+        // We program the enable bits in groups of 32.
+        size_t nSourcesGroups = nSources >> 5;
+
+        auto setField = [&](auto &field, size_t offset, size_t size) {
+          CHERI::Capability capability = root.cast<uint32_t>();
+          capability.address()=  baseAddr + offset;
+          capability.bounds() = size;
+          field               = capability;
+        };
+        setField(plicPrios, PriorityOffset, nSources * sizeof(uint32_t));
+        setField(plicPendings, PendingOffset, nSources / 8);
+        setField(plicEnables, EnableOffset, nSources / 8);
+        setField(plicThres, ThresholdOffset, sizeof(uint32_t));
+        setField(plicClaim, ClaimOffset, sizeof(uint32_t));
+
+        for (size_t i = 0; i < nSourcesGroups; i++) {
+          plicEnables[i] = 0U;
+        }
+
+        for (size_t i = 1; i <= maxIntrID; i++) {
+          plicPrios[i] = 0U;
+        }
+        // We don't make use of threshold control. Leave it at 0 so all
+        // configured interrupts can fire.
+        *plicThres = 0U;
+      }
+
+      /**
+       * Enable the specified interrupt.
+       */
+      void interrupt_enable(Interrupts src) volatile {
```

Comment:
Remove volatile here (and other functions). The `plicEnables` pointer is volatile so the struct doesn't have to be.